468 show captured pieces instead of material imbalance

Jimima opened the pull request:
Had a go at this, draft at this stage as there is a bit to do but wanted feedback on the overall idea, if possible

<img src="https://github.com/user-attachments/assets/5f871f78-2fab-4412-98df-cc77c7d7f8d5" width="300">

Relates https://github.com/lichess-org/mobile/issues/468

veloce commented:
UI looks fine!

Jimima commented:
I added a secondary option to the settings for this. I wonder if this is the best approach, it was simple to add it this way but maybe a single choice with 3 options would be better

<img src="https://github.com/user-attachments/assets/d707dd86-cdc0-4c54-a06d-713606f98b92" width="300">
<img src="https://github.com/user-attachments/assets/6a09aabc-9008-4a4b-aac0-a0fab2ad9017" width="300">

Was also a bit unclear on why the theme preference implementation did something custom for the choice picker on ios, this seems to work but can take the other approach if there is a reason to do so

ijm8710 commented:
I think it makes more sense to have one setting with 3 options. Rather than 2 separate toggle due to the fact that the settings screen keeps getting more options and to limit it best we can.

Obv Veloce/toms opinion will carry more weight but I think it's best to keep that contained best they can

Jimima commented:
Alternate option

<img src="https://github.com/user-attachments/assets/1a9d7660-e54b-4292-b197-7777c8e2980b width="250">
<img src="https://github.com/user-attachments/assets/dfb5f25f-ebe3-40ce-bd11-42845b318930" width="250">

Jimima commented:
In-game option

https://github.com/user-attachments/assets/bf1684f8-4640-40fc-9575-d8cc518bbe88

Jimima commented:
Would appreciate a review @veloce @tom-anders

veloce commented:
It is better with a single setting yes. Thanks!

tom-anders commented on lib/src/model/game/material_diff.dart:
```diff
@@ -80,8 +104,16 @@ class MaterialDiff with _$MaterialDiff {
+      black: MaterialDiffSide(
+        pieces: black.toIMap(),
+        score: -score,
+        capturedPieces: blackCapturedPieces,
```
Can you extend `test/model/game/material_diff_test.dart` with a check for `capturedPieces`?

tom-anders commented on lib/src/model/game/material_diff.dart:
```diff
@@ -35,6 +36,29 @@ class MaterialDiff with _$MaterialDiff {
+      IMap<Role, int> startingCount,
+      IMap<Role, int> subtractCount,
+    ) {
+      IMap<Role, int> capturedPieces = IMap();
```
This could be written in a pure functional way:

```
return startingCount.map(
    (role, count) => MapEntry(role, count - (subtractCount.get(role) ?? 0)),
);

```

Jimima replied:
Thanks, I thought this should be possible but I couldn't figure out the right way to do it

tom-anders commented on lib/src/model/game/material_diff.dart:
```diff
@@ -35,6 +36,29 @@ class MaterialDiff with _$MaterialDiff {
+        Board.standard.materialCount(Side.black);
+    final IMap<Role, int> whiteStartingCount =
+        Board.standard.materialCount(Side.white);
+    // TODO: parameterise starting position maybe so it can be passed in
```
Yes I think this is needed before merging - It's already possible to view (but not play) variants that have a different starting position (e.h. Horde). With this PR, `capturedPieces` will display incorrect values currently.

However, I noticed that the web UI does not display material difference at all for Horde, maybe we should do the same? @veloce wdyt?

tom-anders commented on lib/src/model/settings/board_preferences.dart:
There's some commented-out code here that probably needs to be removed

tom-anders commented on lib/src/model/settings/board_preferences.dart:
```diff
@@ -300,3 +313,17 @@ enum BoardTheme {
+
+  const MaterialDifference({
+    required this.label,
+    required this.visible,
```
is the `visible` field really needed? A property `bool get hidden => this != MaterialDifference.hidden` would also work (and probably same some memory)

tom-anders commented on lib/src/model/settings/board_preferences.dart:
```diff
@@ -300,3 +313,17 @@ enum BoardTheme {
+    required this.visible,
+  });
+
+  final String label;
```
The label will be translated at some point, so it's maybe better to already make this a method along the lines of `String l10n(BuildContext context) => switch (this) { ... }` instead (and add a `//TODO l10n`)

Jimima replied:
Sorry, I am not 100% sure what you mean here (though I understand the broader point). Is there an example somewhere else I could look at?

tom-anders replied:
Sure, I meant something like this: https://github.com/lichess-org/mobile/blob/ca78193fdd972e9e21ffd282cde7ef2642450d74/lib/src/model/study/study_filter.dart#L16

But actually it's more common in the codebase to do it via a free function, like this: https://github.com/lichess-org/mobile/blob/ca78193fdd972e9e21ffd282cde7ef2642450d74/lib/src/view/settings/theme_screen.dart#L31

Jimima replied:
Hmm yeah I think maybe it sitting in the enum itself makes more sense but I'm not sure. I can implement either approach if there is an opinion either way. For now I might just go with option A, it's easy to change

tom-anders commented on lib/src/model/settings/board_preferences.dart:
```diff
@@ -110,7 +119,9 @@ class BoardPrefs with _$BoardPrefs implements Serializable {
     required bool pieceAnimation,
-    required bool showMaterialDifference,
+    // required bool showMaterialDifference,
+    required MaterialDifference materialDifference,
```
Calling this `MaterialDifferenceFormat` (like in the commented line below) would be more readable I think. Currently, there's both `MaterialDiff` and `MaterialDifference` which is confusing

tom-anders commented on lib/src/view/game/game_player.dart:
```diff
@@ -140,16 +143,25 @@ class GamePlayer extends StatelessWidget {
-                    size: 13,
-                    color: Colors.grey,
-                  ),
+              if (materialDifference == MaterialDifference.capturedPieces)
```
Maybe it would be more readable to factor out this whole `Row` into a `class MaterialDifferenceDisplay extends StatelessWidget`. (Then you could also precompute the pieces to loop over in a helper variable and use that in the loop, instead of needing to write the loop twice)

Jimima commented:
@tom-anders I pushed up a bunch of changes, would appreciate you taking another look when you have the time 🙂

tom-anders commented:
very nice, no more comments from me :)

Jimima commented:
Oh should have said, I still don't really know the best label to use for the setting. I just changed it to simply "Material" just now but I've changed it a few times and still can't decide what is best. Happy to be told here!

Jimima commented:
> very nice, no more comments from me :)

Thanks, will take a look at the l10n stuff tomorrow

Jimima commented:
Just in case anyone was wondering I changed the base branch as the diffs were out of date and this was suggested as a fix. Maybe I misunderstood what I was looking at but it seems to have worked so...

Jimima commented:
@veloce @tom-anders I think I've gone as far as I can with this. Would appreciate any further feedback you have for me 🙂

veloce commented on lib/src/view/settings/board_settings_screen.dart:
```diff
@@ -247,13 +248,42 @@ class _Body extends ConsumerWidget {
+                  } else {
+                    pushPlatformRoute(
+                      context,
+                      title: 'Clock position',
```
Wrong title

veloce commented on lib/src/model/settings/board_preferences.dart:
```diff
@@ -120,7 +124,7 @@ class BoardPrefs with _$BoardPrefs implements Serializable {
     required bool coordinates,
     required bool pieceAnimation,
-    required bool showMaterialDifference,
+    required MaterialDifferenceFormat materialDifferenceFormat,
```
A new `required` need a `@JsonKey(defaultValue:` annotation otherwise json parsing fail.

veloce commented:
Thanks for this! I fixed the small issues mentioned above, the rest was perfect.

Jimima commented:
Ah okay that's great, thanks!
